### PR TITLE
GH-40991: [R] Prefer r-universe, add a startup message

### PR DESCRIPTION
**Pull request by jonkeane**
### Rationale for this change

If someone loads a version of Arrow on macOS with features disabled, warn them on startup that they can use `install_arrow()`. By default, prefer R-Universe in `install_arrow()` 

### What changes are included in this PR?

^^^

### Are these changes tested?

Yes

### Are there any user-facing changes?

Yes

**This PR contains a "Critical Fix".**
* GitHub Issue: #40991

### CODE REVIEW

File: r/DESCRIPTION [hunk 1]
```diff
@@ -43,7 +43,7 @@ Imports:
     utils,
     vctrs
 Roxygen: list(markdown = TRUE, r6 = FALSE, load = "source")
-RoxygenNote: 7.2.3
+RoxygenNote: 7.3.1
```
Comment (jonkeane) [inline r/DESCRIPTION:46]: I'm happy to back this out + the extra docs changes, but they were stale on main.

File: r/R/arrow-package.R [hunk 2]
```diff
@@ -182,37 +182,50 @@ configure_tzdb <- function() {
...
-        paste(
-          "Warning:",
-          "  It appears that you are running R and Arrow in emulation (i.e. you're",
-          "  running an Intel version of R on a non-Intel mac). This configuration is",
-          "  not supported by arrow, you should install a native (arm64) build of R",
-          "  and use arrow with that. See https://cran.r-project.org/bin/macosx/",
-          "",
-          sep = "\n"
+  try(
+    {
```
Comment (jonkeane) [inline r/R/arrow-package.R:186]: styler apparently demands this now?. I can back it out if we don't like it, but came from `make doc` so we'll be fighting it if we do.

File: r/R/arrow-package.R [hunk 3]
```diff
@@ -182,37 +182,50 @@ configure_tzdb <- function() {
...
+            "Some features are not enabled in this build of Arrow.",
+            "Run `arrow_info()` for more information."
+          )
         )
-      )
-    }
-  }, silent = TRUE)
+        # On macOS binaries from CRAN can be hobbled. If on macOS, and features
+        # are disabled, advise that reinstalling might help
+        if (identical(tolower(Sys.info()[["sysname"]]), "darwin")) {
```
Comment (jonkeane) [inline r/R/arrow-package.R:217]: We could also check and confirm that the package was installed from CRAN here too if we wanted to be extra sure. Though we will also need to confirm that Posit Public Package Manager isn't also hosting bad binaries too if we do that, but if we leave this as a check on all macs, we have a slightly less strong push to confirm all non-CRANs are building on their own

File: r/R/install-arrow.R [hunk 4]
```diff
@@ -60,6 +62,7 @@ install_arrow <- function(nightly = FALSE,
                           minimal = Sys.getenv("LIBARROW_MINIMAL", FALSE),
                           verbose = Sys.getenv("ARROW_R_DEV", FALSE),
                           repos = getOption("repos"),
+                          prefer_runiverse = TRUE,
```
Comment (jonkeane) [inline r/R/install-arrow.R:65]: I don't _love_ the option profusion here, but I do think going to r-universe first here is actually a nice plan. We don't have to muck with swapping to only source builds on macOS in order to get libarrow (and then also if someone has a borked gcc setup that would error too).

File: r/R/install-arrow.R [hunk 5]
```diff
@@ -76,22 +79,15 @@ install_arrow <- function(nightly = FALSE,
...
-    }
-    if (apple_m1 || rosetta) {
       Sys.setenv(FORCE_BUNDLED_BUILD = "true")
     }
 
     opts <- list()
-    if (apple_m1 || rosetta) {
-      # Skip binaries (esp. for rosetta)
-      opts$pkgType <- "source"
-    } else if (isTRUE(binary)) {
```
Comment (jonkeane) [inline r/R/install-arrow.R:94]: I believe the m1 detection isn't needed any more now that we have binaries for arm. The rosetta checking for jemalloc also probably can go away, since rosetta generally doesn't work (I'm happy to delete it too)

File: r/R/arrow-package.R [hunk 6]
```diff
@@ -182,37 +182,50 @@ configure_tzdb <- function() {
...
+          packageStartupMessage(
+            paste(
+              "The repository you retrieved Arrow from did not include all of Arrow's features.",
+              "You can install a fully-featured version by running `arrow::install_arrow()`."
```
Comment (nealrichardson) [inline r/R/arrow-package.R:221]: Did you see this R-package-devel thread? https://stat.ethz.ch/pipermail/r-package-devel/2024q2/010624.html

TL;DR order of `repos` doesn't necessarily guarantee order of lookup. So it might be better here to just recommend `install.packages("arrow", repos = "https://apache.r-universe.dev")`. We know they already have the arrow package dependencies installed since they're loading it, so we don't need to keep on a CRAN in order to get dependencies to resolve.

File: r/R/install-arrow.R (hunk 4 above)
Comment (nealrichardson) [reply r/R/install-arrow.R:65]: If you take my other suggestion, you don't need to touch this function (though cleaning up the autobrew references is probably worthwhile anyway) or `arrow_repos()`.

File: r/R/arrow-package.R [hunk 7]
```diff
@@ -182,37 +182,50 @@ configure_tzdb <- function() {
...
+        packageStartupMessage(
+          paste(
+            "Some features are not enabled in this build of Arrow.",
+            "Run `arrow_info()` for more information."
+          )
         )
-      )
-    }
-  }, silent = TRUE)
+        # On macOS binaries from CRAN can be hobbled. If on macOS, and features
```
Comment (nealrichardson) [inline r/R/arrow-package.R:215]: I don't think it's a bad idea to explain here why they're hobbled, just so we'll know in the future when we can remove this workaround.

**Review by nealrichardson — COMMENTED** [review]
Thanks for doing this. One suggestion that might simplify things a bit.

File: r/R/arrow-package.R (hunk 6 above)
Comment (jonkeane) [reply r/R/arrow-package.R:221]: Ah right of course. Thanks!

File: r/R/arrow-info.R [hunk 8]
```diff
@@ -139,7 +139,8 @@ arrow_with_json <- function() {
 some_features_are_off <- function(features) {
   # `features` is a named logical vector (as in arrow_info()$capabilities)
   # Let's exclude some less relevant ones
-  blocklist <- c("lzo", "bz2", "brotli", "substrait")
+  # jemalloc is only included because it is sometimes disabled in our build process
+  blocklist <- c("lzo", "bz2", "brotli", "substrait", "jemalloc")
```
Comment (jonkeane) [inline r/R/arrow-info.R:143]: This might be a little paranoid / silly, but we disable jemalloc in our own builds, and honestly if only jemalloc is off, but nothing else is, that's generally ok, so I would rather be less noisy so long as folks can read parquet

File: r/R/install-arrow.R [hunk 9]
```diff
@@ -76,22 +76,15 @@ install_arrow <- function(nightly = FALSE,
       ARROW_R_DEV = verbose,
       ARROW_USE_PKG_CONFIG = use_system
     )
-    # On the M1, we can't use the usual autobrew, which pulls Intel dependencies
-    apple_m1 <- grepl("arm-apple|aarch64.*darwin", R.Version()$platform)
     # On Rosetta, we have to build without JEMALLOC, so we also can't autobrew
```
Comment (nealrichardson) [inline r/R/install-arrow.R:79]:
```suggestion
    # On Rosetta, we have to build without JEMALLOC
```

File: r/R/install-arrow.R [hunk 10]
```diff
@@ -76,22 +76,15 @@ install_arrow <- function(nightly = FALSE,
...
     rosetta <- on_rosetta()
     if (rosetta) {
       Sys.setenv(ARROW_JEMALLOC = "OFF")
-    }
-    if (apple_m1 || rosetta) {
       Sys.setenv(FORCE_BUNDLED_BUILD = "true")
```
Comment (nealrichardson) [inline r/R/install-arrow.R:83]: This is probably also about autobrew right?

```suggestion
```

File: r/R/install-arrow.R [hunk 11]
```diff
@@ -76,22 +76,15 @@ install_arrow <- function(nightly = FALSE,
       ARROW_R_DEV = verbose,
       ARROW_USE_PKG_CONFIG = use_system
     )
-    # On the M1, we can't use the usual autobrew, which pulls Intel dependencies
-    apple_m1 <- grepl("arm-apple|aarch64.*darwin", R.Version()$platform)
     # On Rosetta, we have to build without JEMALLOC, so we also can't autobrew
     rosetta <- on_rosetta()
     if (rosetta) {
```
Comment (nealrichardson) [inline r/R/install-arrow.R:81]:
```suggestion
    if (on_rosetta()) {
```

**Review by nealrichardson — APPROVED** [review]
Some final cleanup suggestions but otherwise LGTM. 

Also need update the PR description to reflect the current scope of the changes.

**Review by thisisnic — APPROVED** [review]
Haven't reviewed in detail but +1 overall to doing this and getting it merged ahead of Monday's code freeze

File: r/R/install-arrow.R (hunk 10 above)
Comment (jonkeane) [reply r/R/install-arrow.R:83]: I'm pretty sure it is, but when I git blamed it wasn't clear that's the only reason for it. I can pull it out, but also we probably should just remove all of the rosetta building code entirely at some point since it's not really a supported setup. I'll add an issue to come back to this (+ various other places we have it)

Comment (jonkeane) [reply r/R/install-arrow.R:82]: #41028
